doc/subsystems/jinjagen/jinjagen.rst

```diff
@@ -0,0 +1,223 @@
+
+Filters can be added to the jinja2 engine by defining custom filters in the python
+script that call the jinja2 engine. This way can functionality easily be added to 
+the jinja2 engine to tailer it for any particular application.
+
+There is no square root filter built into jinja, this can be added by using 
+the ``sqrt()`` function in python to create a jinja2 filter
+
+Python code:
+
+.. code-block:: python
+
+        def squarerootfilter(value):
+                return sqrt(value)
+
+        environment.filters['sqrt'] = squarerootfilter
+
+Jinja2 template:
+
+.. code-block:: c
+        
+        Square root of 5: {{ 5|sqrt() }}
+
+
+
+
+Extending and reusing templates
+===============================
+
+Templates can extend and/or reuse each other through multiple mechanisms. To be able
+extend or reuse a template, the template must be stored in one of the folders listed
+in the search path list. Search paths can be set using the SEARCH_PATH key word in the
+CMakeLists.txt file.
+
+Include
+-------
+
+Include works more or less like include pre-processor directive in C, and will
+insert the content in another template starting from the location of the include keyword.
+
+.. code-block:: c
+
+	{% include 'another_template.jinja2' %}
+
+
+Import
+------
+
+The import keyword will import all macros defined in another template into a namespace.
+Similar to how import in python works. 
+
+.. code-block:: c
+
+	{% import 'template_with_macros.jinja2' as macros %}
+
+	{{ macros.an_imported_macro() }}
+
+
+Extends
+-------
+
+The extends keyword is used to extend and overload macros from another template. 
+Similar to how a child class can do the same with a parent class in object orient programming.
+
+.. code-block:: c
+
+	{% extends 'base_template.jinja2' %}
+
+	{% macro added_macro() %}
+	  This macro is from the child template
+	{% endmacro %}
+
+	{% block overload_content_in_parent_template %}
+	  This is inserted into the parent template class block from the child template
+	{% endblock %}
+
+Block
+^^^^^
+
+The block keyword can be used to define a block in the parent template that can be overloaded
+by the child template. The child block can also include the content parent block by calling
+super().
+
+.. code-block:: c
+
+    {# This is the parent template #}
+
+    {% block a_block %}
+      This is the content of the parent block
+    {% endblock %}
+
+    {# This is the child template #}
+    {% extends 'parent_template.jinja2' %}
+
+    {% block  a_block() %}
+       This is the text from the parent block: {{ super() }}
+       I add some text here!
+    {% endblock %}
+
+Code Generation in the Build Process
+====================================
+
+Code generation has to be invoked as part of the build process. Zephyr uses
+`CMake <https://cmake.org/>`_ as the tool to manage building the project.
+
+A template file that must be rendered before it can be used in a project 
+by one of the following commands in a `CMakeList.txt` file.
```

Comment:
A template file can be inserted into the build by one of the following CMake functions.

Reply:
ok